apollo-router/src/plugins/healthcheck/mod.rs

```diff
@@ -0,0 +1,506 @@
+//! Health Check plugin
+//!
+//! Provides liveness and readiness checks for the router.
+//!
+//! This module needs to be executed prior to traffic shaping so that it can capture the responses
+//! of requests which have been load shed.
+//!
+
+use std::net::SocketAddr;
+use std::str::FromStr;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::time::Duration;
+
+use http::StatusCode;
+use multimap::MultiMap;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use tokio::time::Instant;
+use tower::service_fn;
+use tower::BoxError;
+use tower::ServiceBuilder;
+use tower::ServiceExt;
+
+use crate::configuration::ListenAddr;
+use crate::plugin::PluginInit;
+use crate::plugin::PluginPrivate;
+use crate::register_private_plugin;
+use crate::services::router;
+use crate::Endpoint;
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "UPPERCASE")]
+#[allow(dead_code)]
+enum HealthStatus {
+    Up,
+    Down,
+}
+
+#[derive(Debug, Serialize)]
+struct Health {
+    status: HealthStatus,
+}
+
+/// Configuration options pertaining to the readiness health interval sub-component.
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+#[serde(default)]
+pub(crate) struct ReadinessIntervalConfig {
+    #[serde(deserialize_with = "humantime_serde::deserialize", default)]
+    #[serde(serialize_with = "humantime_serde::serialize")]
+    #[schemars(with = "Option<String>", default)]
+    /// The sampling interval (default: 10s)
+    pub(crate) sampling: Duration,
+
+    #[serde(deserialize_with = "humantime_serde::deserialize")]
+    #[serde(serialize_with = "humantime_serde::serialize")]
+    #[schemars(with = "Option<String>")]
+    /// The unready interval (default: 3 * sampling interval)
+    pub(crate) unready: Option<Duration>,
+}
+
+/// Configuration options pertaining to the readiness health sub-component.
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+#[serde(default)]
+pub(crate) struct ReadinessConfig {
+    /// The readiness interval configuration
+    pub(crate) interval: ReadinessIntervalConfig,
+
+    /// How many errors/interval are allowed until unready (default: 100)
```

Comment:
is `100` the _number_ of errors, or 100 second interval? the `errors/interval` is slightly misleading to me.

Reply:
It is the number of rejections per interval. I'll phrase that more clearly.